### DISCUSSION TITLE
[java rules]How to get param's class?

### DISCUSSION THREAD

**Discussion by JiayuK** [pmd/pmd#5242] [Q&A] [answered]
Please help me, I need to check if instance a uses methods containing special characters, if this class extends from the specified class.
for example

```java
public class A extends Aa{}

public class main {
    public static void main(String[] args) {
        A a = new A();
       setDump(a);
    }
}
```
I don't want to extends the instance of Aa and set it to dump.

My current implementation method,

```java
public class Dump extends AbstractJavaRule {
    @Override
    public Object visit(ASTExpressionStatement node, Object data) {
        Node node1 = node.getFirstChild();
        if (node1.getImage().contains("setDump")) {
        }
        return super.visit(node, data);
    }
```

Question 1: I don't know how to traverse and find the VariableAccess of ArgumentList
Question 2: The Variable Access I obtained through traversal is of type Node, and I don't know how to obtain the class corresponding to the variable

![6CF64D54-2852-40E4-BF05-6872040A99B7](https://github.com/user-attachments/assets/b3e69d27-b6a7-42e5-9569-44f53705ba7c)

Please help me, thanks

**Comment by jsotuyod** [comment, marked as answer]
1. You probably want to do something such as
  ```java
if (node1 instanceof ASTMethodCall && node1.getImage().contains("setDump")) {
  ASTArgumentList args = ((ASTMethodCall) node1).getArguments();
  // args is iterable, and can be accessed pretty much like a list… ie: args.get(0);
  // you don't really care about they type of node (ie: VariableAccess vs FieldAccess), just that it's a TypeNode…
}
  ```
2. You can use any of the overloads in `TypeTestUtil.isA` to check the type of a `TypeNode`, even considering inheritance.

**Reply by JiayuK** [reply]
Thank you for your reply
May I ask what method I can use to obtain the VariableAccess of MethodCall
![image](https://github.com/user-attachments/assets/db07f5d8-c797-4467-a3f2-f78d4cea281e)

**Reply by jsotuyod** [reply]
The complete documentation on node transversal is available [here](https://docs.pmd-code.org/latest/pmd_userdocs_extending_writing_java_rules.html#manual-ast-navigation)
